src/distro_release.rs

```diff
@@ -0,0 +1,274 @@
+    eol_server: Option<NaiveDate>,
+}
+
+impl DistroRelease {
+    pub fn new(
+        version: String,
+        codename: String,
+        series: String,
+        created: Option<NaiveDate>,
+        release: Option<NaiveDate>,
+        eol: Option<NaiveDate>,
+        eol_lts: Option<NaiveDate>,
+        eol_elts: Option<NaiveDate>,
+        eol_esm: Option<NaiveDate>,
+        eol_server: Option<NaiveDate>,
+    ) -> Self {
+        Self {
+            version: if version.is_empty() {
+                None
+            } else {
+                Some(version)
+            },
+            codename,
+            series,
+            created,
+            release,
+            eol,
+            eol_lts,
+            eol_elts,
+            eol_esm,
+            eol_server,
+        }
+    }
+
+    // Getters
+    pub fn version(&self) -> &Option<String> {
+        &self.version
+    }
+    pub fn codename(&self) -> &String {
+        &self.codename
+    }
+    pub fn series(&self) -> &String {
+        &self.series
+    }
+    pub fn created(&self) -> &Option<NaiveDate> {
+        &self.created
+    }
+    pub fn release(&self) -> &Option<NaiveDate> {
+        &self.release
+    }
+    pub fn eol(&self) -> &Option<NaiveDate> {
+        &self.eol
+    }
+    pub fn eol_server(&self) -> &Option<NaiveDate> {
+        &self.eol_server
+    }
+    pub fn eol_esm(&self) -> &Option<NaiveDate> {
+        &self.eol_esm
+    }
+    pub fn eol_elts(&self) -> &Option<NaiveDate> {
+        &self.eol_elts
+    }
+    pub fn eol_lts(&self) -> &Option<NaiveDate> {
+        &self.eol_lts
+    }
+
+    // Non-getters
+    pub fn ubuntu_is_lts(&self) -> bool {
+        self.version
+            .as_ref()
+            .map(|version| version.contains("LTS"))
+            .unwrap_or(false)
+    }
+
+    pub fn created_at(&self, date: NaiveDate) -> bool {
+        match self.created {
+            Some(created) => date >= created,
+            None => false,
+        }
+    }
+
+    pub fn released_at(&self, date: NaiveDate) -> bool {
+        match self.release {
+            Some(release) => date >= release,
+            None => false,
+        }
+    }
+
+    pub fn milestone_date(&self, milestone: &Milestone) -> Option<NaiveDate> {
+        *match milestone {
+            Milestone::Eol => self.eol(),
+            Milestone::EolELTS => self.eol_elts(),
+            Milestone::EolESM => self.eol_esm(),
+            Milestone::EolLTS => self.eol_lts(),
+            Milestone::EolServer => self.eol_server(),
+        }
+    }
+
+    pub fn supported_at(&self, date: NaiveDate, milestone: &Milestone) -> bool {
+        self.created_at(date)
+            && match self.milestone_date(milestone) {
+                Some(eol) => date <= eol,
+                // Missing eol means supported, otherwise unsupported
+                None => milestone == &Milestone::Eol,
+            }
+    }
+
+    pub fn ubuntu_supported_at(&self, date: NaiveDate) -> bool {
```

Comment:
it seems a bit odd to have this on the trait rather than just UbuntuDistroInfo; when would you want to call this on a DebianDistroInfo object?

Reply:
Thanks for the review! This is part of `DistroRelease`'s implementation (which represents individual items) rather than `DistroInfo` (which is the trait implemented for `{Debian,Ubuntu}DistroInfo`).

I agree that it's strange to have this here, but we don't currently have a distinct `UbuntuDistroRelease` type, and the APIs above all return `DistroRelease` (or a `Vec<_>` thereof) and only call the distro-specific methods when appropriate. It would definitely be good to have specialised structs to avoid the need to "know" which distro you're dealing with, but I think a fair amount of refactoring would be involved.